src/interface/sparse_pack_base.hpp

```diff
@@ -0,0 +1,199 @@
+//========================================================================================
+// (C) (or copyright) 2020-2022. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_SPARSE_PACK_BASE_HPP_
+#define INTERFACE_SPARSE_PACK_BASE_HPP_
+
+#include <algorithm>
+#include <functional>
+#include <limits>
+#include <map>
+#include <memory>
+#include <regex>
+#include <string>
+#include <tuple>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "coordinates/coordinates.hpp"
+#include "interface/variable.hpp"
+#include "utils/utils.hpp"
+
+namespace parthenon {
+namespace impl {
+struct PackDescriptor {
```

Comment:
This is nice. Maybe we should use something like this for the whole non-sparse pack machinery. Or is the idea that `SparsePack` eventually replaces dense packs?

Reply:
I am not sure about the best way forward. I think `SparsePack`s have essentially all of the same functionality of the pre-existing packing machinery, but having them be the only packs would mean that a lot of other code needs to be changed.

Reply:
I lean towards replacing the old packs completely. But agreed htat that's a surgery. Once this is all merged in I think we should have a separate PR that does this, which probably also implies a code version bump. Thoughts, @jdolence @forrestglines @pgrete ?

Reply:
Yeah, we also need to do performance testing as well before making such a big change I think